### PR TITLE
Create MXSecretStorage to support SSSS

### PR DESCRIPTION
**Pull request by manuroe**
Fix https://github.com/vector-im/riot-ios/issues/2926

Spec: ([MSC1946(]https://github.com/matrix-org/matrix-doc/pull/1946)

This module will be used for bootstrap and account keys recovery.

image

### CODE REVIEW

File: MatrixSDK/Crypto/SecretStorage/MXSecretStorage.m [hunk 1]
```diff
@@ -0,0 +1,574 @@
...
+    return key;
+}
+
+- (void)checkPrivateKey:(NSData *)privateKey withKey:(MXSecretStorageKeyContent *)key complete:(void (^)(BOOL match))complete
+{
+    MXWeakify(self);
+    dispatch_async(processingQueue, ^{
+        MXStrongifyAndReturnIfNil(self);
+        
+        NSData *iv = [MXBase64Tools dataFromBase64:key.iv];
```
Comment (SBiOSoftWhare) [inline MatrixSDK/Crypto/SecretStorage/MXSecretStorage.m:183]: `MXSecretStorageKeyContent.iv` is nullable. Crash can occurs here.

File: MatrixSDK/Crypto/SecretStorage/MXSecretStorage.m [hunk 2]
```diff
@@ -0,0 +1,574 @@
...
+        
+        // MACs should match
+        NSError *error;
+        MXEncryptedSecretContent *encryptedZeroString = [self encryptedZeroStringWithPrivateKey:privateKey iv:iv error:&error];
+        
+        // Compare bytes instead of base64 strings to avoid base64 padding issue
+        NSData *keyMac = [MXBase64Tools dataFromBase64:key.mac];
+        NSData *encryptedZeroStringMac = [MXBase64Tools dataFromBase64:encryptedZeroString.mac];
```
Comment (SBiOSoftWhare) [inline MatrixSDK/Crypto/SecretStorage/MXSecretStorage.m:191]: MXSecretStorageKeyContent.mac is nullable. Crash can occurs here.

File: MatrixSDK/Crypto/SecretStorage/MXSecretStorage.m [hunk 3]
```diff
@@ -0,0 +1,574 @@
...
+                                            outputLength:64];
+    
+    // The first 32 bytes are used as the AES key, and the next 32 bytes are used as the MAC key
+    NSData *aesKey = [pseudoRandomKey subdataWithRange:NSMakeRange(0, 32)];
+    NSData *hmacKey = [pseudoRandomKey subdataWithRange:NSMakeRange(32, pseudoRandomKey.length - 32)];
+
+
+    NSData *iv = secretContent.iv ? [MXBase64Tools dataFromBase64:secretContent.iv] : [NSMutableData dataWithLength:16];
+    
+    NSData *hmac = [MXBase64Tools dataFromBase64:secretContent.mac];
```
Comment (SBiOSoftWhare) [inline MatrixSDK/Crypto/SecretStorage/MXSecretStorage.m:536]: MXEncryptedSecretContent.mac is nullable. Crash can occurs here.

File: MatrixSDK/Crypto/SecretStorage/MXSecretStorage.m [hunk 4]
```diff
@@ -0,0 +1,574 @@
...
+    if (!hmac)
+    {
+        NSLog(@"[MXSecretStorage] decryptSecret: ERROR: Bad base64 format for MAC: %@", secretContent.mac);
+        *error = [self errorWithCode:MXSecretStorageBadMacCode reason:[NSString stringWithFormat:@"Bad base64 format for MAC: %@", secretContent.mac]];
+        return nil;
+    }
+
+    NSData *cipher = [MXBase64Tools dataFromBase64:secretContent.ciphertext];
```
Comment (SBiOSoftWhare) [inline MatrixSDK/Crypto/SecretStorage/MXSecretStorage.m:544]: MXEncryptedSecretContent.cipherText is nullable. Crash can occurs here.

File: MatrixSDK/Crypto/SecretStorage/MXSecretStorage.h [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+ 
+ @param secretId the id of the secret.
+ @param keyId the id of the key to use to decrypt it. Use secretStorageKeysUsedForSecretWithSecretId to get it.
+              Nil will use the default key.
+ @param privateKey the private key for this key.
+ @param success A block object called when the operation succeeds.
+ @param failure A block object called when the operation fails.
+ */
+- (void)secretWithSecretId:(NSString*)secretId
+    withSecretStorageKeyId:(nullable NSString*)keyId
```
[the remark this answers is not in the document]
Comment (manuroe) [reply MatrixSDK/Crypto/SecretStorage/MXSecretStorage.h:144]: why? SecretId and SecretStorageKeyId are 2 different things.

[the remark this answers is not in the document]
Comment (SBiOSoftWhare) [reply MatrixSDK/Crypto/SecretStorage/MXSecretStorage.h:143]: Only the second one

**Pull request merged**